pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintService.kt

```diff
@@ -25,22 +26,32 @@ class PrintService @Inject constructor(private var factory: PrinterFactory) {
     try {
       val dto = request.read(ReceiptDTO::class.java).`as`(GsonTransport::class.java)
       val printer = factory.getPrinter(dto.sourceIp)
-      when {
-        dto.fiscal
-        -> response = printer.printFiscalReceipt(dto.receipt)
-        else
-        -> response = printer.printReceipt(dto.receipt)
+      try {
+        when {
+          dto.fiscal
+          -> response = printer.printFiscalReceipt(dto.receipt)
+          else
+          -> response = printer.printReceipt(dto.receipt)
+        }
+      } catch (e: RequestTimeoutException) {
+        return Reply.with(ErrorResponse("Printer request timeout.\n" + e.message)).`as`(GsonTransport::class.java).status(504)
+      } finally {
+        printer.close()
       }
-      printer.close()
     } catch (e: DeviceNotFoundException) {
-      return Reply.with(ErrorResponse("Device not found.")).badRequest()
+      return Reply.with(ErrorResponse("Device not found.")).`as`(GsonTransport::class.java).notFound()
     } catch (e: IOException) {
-      return Reply.with(ErrorResponse("Device can't connect.")).status(480)
+      return Reply.with(ErrorResponse("Device can't connect.")).`as`(GsonTransport::class.java).status(480)
     }
+
     val responseDTO: PrintReceiptResponseDTO = PrintReceiptResponseDTO(response.warnings)
-    return Reply.with(responseDTO).`as`(GsonTransport::class.java)
-  }
+    println(GsonBuilder().create().toJson(responseDTO))
 
+    return if (responseDTO.warnings.contains(Status.FISCAL_RECEIPT_IS_OPEN) ||
+      responseDTO.warnings.contains(Status.NON_FISCAL_RECEIPT_IS_OPEN)) {
+      Reply.with(responseDTO).`as`(GsonTransport::class.java)
+    } else Reply.with(ErrorResponse("Didn't open receipt.")).`as`(GsonTransport::class.java).badRequest()
```

Comment:
It means that there is a connection and the printer received the request, but didn't open a receipt.
So it must be safe to try again printing the receipt. I think is better to remove it, but warnings should be checked somewhere else, because the reason for not opening receipt in this case should be explained by a warning like:
END_OF_PAPER, BROKEN_PRINTIN_MECHANISM, COVER_IS_OPEN.